java generation classes rewrited to kotlin

pakhopav opened the pull request.

pakhopav committed:
java generation classes rewrited to kotlin

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/BnfGenerator.kt:
```diff
@@ -0,0 +1,247 @@
+  generateTokenAccessors=true
+  generateTokens=true
+
+}
```
подозрительно, лучше структурировать так, чтобы открывающие скобки генерились тем же кодом что закрывающие

а то непонятно, где ее пара

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/BnfGenerator.kt:
```diff
@@ -0,0 +1,247 @@
+        for (b in branches) {
+            writebrunch(b, sb)
+            sb.append("${if (b != branches.last()) {
+                "| "
```
use join instead

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/EnumRule.kt:
```diff
@@ -0,0 +1,8 @@
+
+import com.intellij.xtextLanguage.xtext.psi.XtextEnumRule
+
+class EnumRule(val myRule: XtextEnumRule) {
```
в котлине не приняты my-префиксы, `myXxx -> xxx`, как ни больно мне это признавать

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/XtextFileModel.java:
```diff
@@ -4,74 +4,120 @@
+                .findFirst().orElse(null);
     }
 
+    private void registerReferences(XtextParserRule rule) {
```
очень подозрительный код
выброси все проверки на `size == 0` -- в том и смысл никогда не возвращать null что не надо проверять, потому что пустой лист можно проитерировать без проблем

borlander replied:
лучше наверное переписать в java9-стримы или на котлин с использованием магического слова FLATTEN

borlander replied:
когда итерация такая большая, лучше разнести итерацию от конечной обработки - выдели конечную обработку в отдельный метод

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/keywords/KeywordCompletionProvider.java:
```diff
@@ -7,9 +7,12 @@
 import com.intellij.codeInsight.completion.CompletionResultSet;
```
кажется, тоже несвязанный коммит

pakhopav committed:
test for negated terminals added

pakhopav committed:
negated terminals generation
tests passed

pakhopav committed:
enum rules generation

pakhopav committed:
method for creating parser rules from text
gen folder regenerated

pakhopav committed:
action support added
all helper methods grom BnfGenerator replaced to BnfGeneratorUtil class

pakhopav committed:
generation imported grammars
temporary implemented by method BuildModelWithImports() in XtextGenerateBnfTestbase class
2 bugs fixed (generation predicated keywords and references registration in Model)
regenerated gen dir without changes
1 pseudo-test for generation imported grammars added

pakhopav committed:
generation bugs fixed
'^' symbol in IDs changed to "caret"
repeating references bug

pakhopav committed:
tests for generation
xtext library files added to test resources

pakhopav committed:
namedElement and abstractRule classes deleted
classes working with them temporary commented

pakhopav committed:
tokens for keyword generation
last changes before bootstrap

pakhopav committed:
bootstrap
parser files regenerated from generated bnf

pakhopav committed:
generation all files for supporting new languages

pakhopav committed:
new testing grammar with expected NameVisitor(writed by hands)

pakhopav committed:
testing method for Choices grammar
creation of nameVisitorModel in testing method because it must evaluate after computing imported grammars(which is in testing method for now )

pakhopav committed:
VisitorGeneratorModel interface and visitorGenerator for generation nameVisitor(finding names of the rules)

pakhopav committed:
"named" ParserRules (if reference to this rule exists), visitorGeneratorModele creation method, searching rules by "Model object" name method

pakhopav committed:
Unimportant changes in generator classes

pakhopav committed:
VisitorGeneratorModel implementation

pakhopav committed:
Generation all Files for reference contribution

pakhopav committed:
branching helper classes (same for all grammars)

pakhopav committed:
XtextGrammar regenerated

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/BnfGenerator.kt:
```diff
@@ -0,0 +1,964 @@
+package com.intellij.xtextLanguage.xtext.generator
```
Хорошо бы разбить на кусочки чтобы избежать класса на ХХХХ строк.
Кажется, что каждый из классов можно генерить отдельно

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/BnfGeneratorOld.java:
```diff
@@ -1,82 +1,115 @@
-                anotations[i] = xtextAnnotation.getText();
-                i++;
-            }
+public class BnfGeneratorOld {
```
Не понимаю пока, надо на это смотреть сейчас?

borlander commented on src/main/java/com/intellij/xtextLanguage/xtext/generator/VisitorGenerator.kt:
```diff
@@ -0,0 +1,67 @@
+            out.print("    fun $methodName (node : ${extention}${model.getXtextRuleName(it)}): PsiElement?{\n")
+            model.getUniqueName(it)?.let {
+                out.print("        return node.${it.decapitalize()}\n")
+                out.print("        return null\n")
```
подозрительно

pakhopav committed:
xtext visitor fot visitorGeneratorModel
doesnt work with actions

pakhopav committed:
3 visitors and one superClass(XtextVisitorRuleCalls) for generation nameVisitor(finding names of the rules)

pakhopav committed:
new VisitorGenerationModel (works with visitors now)
